src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsProjectJsonToPackageReferenceMigrator.cs

```diff
@@ -61,25 +54,42 @@ public IVsProjectJsonToPackageReferenceMigrateResult MigrateProjectJsonToPackage
             var projectSafeName = await EnvDTEProjectUtility.GetCustomUniqueNameAsync(project);
 
             var nuGetProject = _solutionManager.Value.GetNuGetProject(projectSafeName);
+
+            // If the project already has PackageReference, do nothing.
+            if (nuGetProject is LegacyCSProjPackageReferenceProject)
+            {
+                return new VsProjectJsonToPackageReferenceMigrateResult(success: true, errorMessage: null);
+            }
+
             try
             {
+                _solutionManager.Value.SaveProject(nuGetProject);
                 var legacyPackageRefBasedProject = await
                     _solutionManager.Value.UpdateNuGetProjectToPackageRef(nuGetProject) as LegacyCSProjPackageReferenceProject;
 
-                await ProjectJsonToPackageRefMigrator.MigrateAsync(
+                if (legacyPackageRefBasedProject != null)
+                {
+                    await ProjectJsonToPackageRefMigrator.MigrateAsync(
                         legacyPackageRefBasedProject,
                         legacyPackageRefBasedProject.MSBuildProjectPath);
-                var result = new VsProjectJsonToPackageReferenceMigrateResult(success: true, errorMessage: null);
-                _solutionManager.Value.SaveProject(nuGetProject);
+                    var result = new VsProjectJsonToPackageReferenceMigrateResult(success: true, errorMessage: null);
+                    _solutionManager.Value.SaveProject(nuGetProject);
+
+                    return result;
+                }
+                else
+                {
+                    throw new InvalidCastException(string.Format(VsResources.Error_InvalidPackageRefProjectCast, projectUniqueName));
```

Comment:
Perhaps add a code comment here explaining why you use `as` and then throw an `InvalidCastException` message yourself instead of just using the `()` cast operator. I'm guessing it's so you can provide the path to the project in question, which sounds worthwhile.